app/services/checksum_validator.rb

```diff
@@ -1,6 +1,15 @@
 # code for validating Moab checksums
 class ChecksumValidator
   include ::MoabValidationHandler
+  include ActiveModel::Validations
+
+  validates_each :endpoint do |record, attr, value|
+    if value.is_a?(Endpoint)
```

Comment:
Is this object valid if the Endpoint isn't also valid?  Type-checking and cascading validity checks are the kind of thing that an ActiveRecord association would deliver inherently.